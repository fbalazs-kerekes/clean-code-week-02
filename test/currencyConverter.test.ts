import { mock, mockReset } from 'jest-mock-extended';
import { IExchangeRateService } from '../src/exchangeRateService';
import { CurrencyConverter } from '../src/currencyConverter';
import { CurrencyFetchError } from '../src/errors/currencyFetchError';

const mockExchangeRateService = mock<IExchangeRateService>();

describe('CurrencyConverter tests', () => {
    const fromCurrency = 'EUR';
    const toCurrency = 'HUF';
    const amount = 1;
    const exchangeRate = 1;
    const reportStartDate = new Date('2024-01-01');
    const reportEndDate = new Date('2024-01-05');
    const expectedExchangeRateErrorMessage = 'Unable to fetch exchange rate.';
    const expectedAmountErrorMessage = 'Invalid amount input.';
    const expectedExchangeRateError = new CurrencyFetchError(expectedExchangeRateErrorMessage);
    const expectedAmountError = new CurrencyFetchError(expectedAmountErrorMessage);
    let currencyConverter: CurrencyConverter;

    beforeEach(() => {
        mockReset(mockExchangeRateService);
        currencyConverter = new CurrencyConverter(mockExchangeRateService);
    })

    describe('Convert tests', () => {
        describe('Happy path', () => {
            it.each([
                ['EUR', 'HUF', 1.0, 392, 392.0],
                ['HUF', 'EUR', 392.0, 0.00255, 1.0],
            ])('should convert currency %s to %s', (fromCurrency: string, toCurrency: string, amount: number, exchangeRate: number, expectedCovertedAmount: number) => {
                mockExchangeRateService.getExchangeRate.mockReturnValue(exchangeRate);
                
                let convertedAmount = Math.round(currencyConverter.Convert(amount, fromCurrency, toCurrency) * 100) / 100;
                
                expect(convertedAmount).toEqual(expectedCovertedAmount);
                expect(mockExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(1);
            })
        })

        describe('Error path', () => {
            it('should throw and error when amount is not a number', () => {
                const amountNaN = NaN;

                mockExchangeRateService.getExchangeRate.mockReturnValue(exchangeRate);

                expect(() => currencyConverter.Convert(amountNaN, fromCurrency, toCurrency)).toThrow(expectedAmountError);
                expect(mockExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(0);
            })

            it('should throw an Error when exchange rate is 0', () => {
                const exchangeRateReturnValue = 0;

                mockExchangeRateService.getExchangeRate.mockReturnValue(exchangeRateReturnValue);
                expect(() => currencyConverter.Convert(amount, fromCurrency, toCurrency)).toThrow(expectedExchangeRateError);
                expect(mockExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(1);
            })


            it('should throw an Error when exchange rate is not a number', () => {
                const exchangeRateReturnValue = NaN;

                mockExchangeRateService.getExchangeRate.mockReturnValue(exchangeRateReturnValue);
                const sut = () => currencyConverter.Convert(amount, fromCurrency, toCurrency);
                expect(sut).toThrow(CurrencyFetchError);
                expect(sut).toThrow(expectedExchangeRateError);
                expect(mockExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(2);
            })
        })
    })

    describe('GenerateConversionReport tests', () => {
        describe('Happy path', () => {
            it.each([
                ['EUR', 'HUF', new Date('2024-01-01'), new Date('2024-01-05'), [1, 0.99, 0.98, 1, 1]],
                ['HUF', 'EUR', new Date('2024-01-01'), new Date('2024-01-05'), [0.04, 0.02, 0.01, 0.05, 0.08]],
            ])('should generate conversion report for %s to %s between %s and %s', (fromCurrency: string, toCurrency: string, startDate: Date, endDate: Date, conversionRates: number[]) => {
                conversionRates.every((value) => mockExchangeRateService.getExchangeRate.mockReturnValueOnce(value));
                
                let conversionReport = currencyConverter.GenerateConversionReport(fromCurrency, toCurrency, startDate, endDate);

                expect(conversionReport).toMatchSnapshot();
                expect(mockExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(5);
            })
        })

        describe('Error path', () => {
            it('should throw and error when exchange rate is not a number', () => {
                const exchangeRateNaN = NaN;

                mockExchangeRateService.getExchangeRate.mockReturnValue(exchangeRateNaN);

                const sut = () => currencyConverter.GenerateConversionReport(fromCurrency, toCurrency, reportStartDate, reportEndDate);
                expect(sut).toThrow(CurrencyFetchError);
                expect(sut).toThrow(expectedExchangeRateError);

                expect(mockExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(2);
            })
        })
    })
})