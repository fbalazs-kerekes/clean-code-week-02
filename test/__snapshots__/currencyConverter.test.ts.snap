// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter tests GenerateConversionReport tests Happy path should generate conversion report for EUR to HUF between 2024-01-01T00:00:00.000Z and 2024-01-05T00:00:00.000Z 1`] = `
"Conversion Report:
100
99
98
100
100"
`;

exports[`CurrencyConverter tests GenerateConversionReport tests Happy path should generate conversion report for HUF to EUR between 2024-01-01T00:00:00.000Z and 2024-01-05T00:00:00.000Z 1`] = `
"Conversion Report:
4
2
1
5
8"
`;
